AzureBatchQueue/TimerBatch.cs

```diff
@@ -114,29 +121,64 @@ async Task Quarantine()
                 var remaining = Remaining();
                 await batchQueue.QuarantineData(msg.MessageId, remaining);
 
-                logger.LogInformation("Message {msgId} was quarantined after {dequeueCount} unsuccessful attempts in queue {QueueName}." +
-                                      " With {remainingCount} unprocessed from {totalCount} total",
+                logger.LogInformation("Message {MsgId} was quarantined after {DequeueCount} unsuccessful attempts in queue {QueueName}." +
+                                      " {RemainingCount} items were not completed ({NotProcessed} not processed on time and {FailedCount} failed) from {TotalCount} total",
                     msg.MessageId,
                     msg.Metadata.DequeueCount,
                     batchQueue.Name,
-                    remaining,
+                    remaining.Length,
+                    NotProcessedCount(),
+                    FailedCount(),
                     items.Items().Length);
             }
         }
     }
 
-    T[] Remaining() => items.NotEmptyItems().Select(x => x.Item).ToArray();
+    T[] Remaining() => items.Remaining().Select(x => x.Item).ToArray();
+    int NotProcessedCount() => items.NotProcessedCount();
+    int FailedCount() => items.FailedCount();
 
     public void Complete(BatchItemId itemId)
+    {
+        ThrowIfCompleted(itemId);
+
+        var remaining = items.Complete(itemId);
+
+        FlushIfEmpty(remaining);
+    }
+
+    public void Fail(BatchItemId itemId)
+    {
+        ThrowIfCompleted(itemId);
+
+        var remaining = items.Fail(itemId);
+
+        FlushIfEmpty(remaining);
+    }
+
+    public void Delay(BatchItemId itemId, TimeSpan delayTime)
```

Comment:
I am unsure about exposing such a specific API in a general-purpose queue. It smells a bit.
Maybe it's better to implement delays in place where they are needed. I.e. by putting data that need to be delayed as a separate message into the queue.

Imagine a case when we receive a batch of mentions that we want to delay (e.g. some news from the provider). And those mentions have different delay times. We will try to process same queue message several times then (as each time, we will delay the rest of the unprocessed mentions). Eventually, the queue message might go to quarantine because of exceeding the `dequeue count`.

I would leave delay functionality to library clients and try to keep the batch queue as simple as possible (only cover cases that are a must for all the scenarios).

Reply:
I don't think the concept of Delay is a specific case in a general-purpose queue. What we do here is basically use the visibility timeout feature of the AzureQueue. But the implementation of Delay in my case is really not thought well-enough and needs a lot more details, so I will move it to the clients.